src/event_loop.rs

```diff
@@ -374,13 +374,26 @@ impl<T> EventLoop<T>
                         CHANNEL => if !self.channel_event(&mut state) {
                             break 'event_loop;
                         },
+
+                        #[cfg(unix)]
+                        token if token == self.pty.signal_token() => {
+                            let should_exit = self.pty.process_signals();
+                            if should_exit {
+                                let mut term = self.terminal.lock();
+                                term.exit();
+                                self.display.notify();
+                                break 'event_loop;
+                            }
+                        },
+
                         token if token == self.pty.read_token() || token == self.pty.write_token() => {
-                            #[cfg(unix)]
-                                {
-                                    if UnixReady::from(event.readiness()).is_hup() {
-                                        break 'event_loop;
-                                    }
+                            #[cfg(unix)] {
+                                if UnixReady::from(event.readiness()).is_hup() {
+                                    // don't try to do I/O on a dead Pty
+                                    continue;
```

Comment:
Not doing this caused the `pty_read` / `_write` to `EIO`, and I'm still not totally sure I've fixed that.

Reply:
So, it turned out the `EIO` was actually a racy bug that's been present for who-knows-how-long, and it's a *miracle* it hasn't burnt us before - after thinking I'd screwed up somehow for a couple days, I finally reproduced it on 0.2.9. See the comments in the new code below.